sql/logictest: no way to guarantee upgrade completion for mixed-version tests

pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test.TestLogic_mixed_version_udf_mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12238646?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12238646?buildTab=artifacts#/) on master @ [3550748890491a8e4e83589f11d1a5bacc36aede](https://github.com/cockroachdb/cockroach/commits/3550748890491a8e4e83589f11d1a5bacc36aede):


```
=== RUN   TestLogic_mixed_version_udf_mutations
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/9e2b4900e7c24e0e704b56d8bacefc6e/logTestLogic_mixed_version_udf_mutations3809825258
    test_log_scope.go:81: use -show-logs to present logs inline
[16:22:15] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/7950/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations: 8 statements
    logic.go:2906: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/7950/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations:90: SELECT crdb_internal.is_at_least_version('23.1-38')
        expected:
            true
            
        but found (query options: "retry") :
            false
            
[16:23:35] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/7950/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations: 14 statements
[16:23:35] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/7950/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations with config cockroach-go-testserver-upgrade-to-master: 14 tests, 1 failures
    logic.go:3996: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/7950/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations:95: error while processing
    logic.go:3996: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/7950/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations:95: too many errors encountered, skipping the rest of the input
cockroach logs captured in: /artifacts/tmp/_tmp/9e2b4900e7c24e0e704b56d8bacefc6e/cockroach-logs3140664564
    panic.go:522: -- test log scope end --
--- FAIL: TestLogic_mixed_version_udf_mutations (116.24s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_mixed_version_udf_mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>



Jira issue: CRDB-32522

pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test.TestLogic_mixed_version_udf_mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12265376?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12265376?buildTab=artifacts#/) on master @ [40c1ff68bbbe3e5f123c9147c3d43796c2935f55](https://github.com/cockroachdb/cockroach/commits/40c1ff68bbbe3e5f123c9147c3d43796c2935f55):


```
=== RUN   TestLogic_mixed_version_udf_mutations
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/9e2b4900e7c24e0e704b56d8bacefc6e/logTestLogic_mixed_version_udf_mutations147623490
    test_log_scope.go:81: use -show-logs to present logs inline
[05:38:00] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/8932/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations: 8 statements
    logic.go:2908: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/8932/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations:90: SELECT crdb_internal.is_at_least_version('23.1-38')
        expected:
            true
            
        but found (query options: "retry") :
            false
            
[05:39:20] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/8932/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations: 14 statements
[05:39:20] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/8932/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations with config cockroach-go-testserver-upgrade-to-master: 14 tests, 1 failures
    logic.go:3998: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/8932/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations:95: error while processing
    logic.go:3998: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/8932/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations:95: too many errors encountered, skipping the rest of the input
cockroach logs captured in: /artifacts/tmp/_tmp/9e2b4900e7c24e0e704b56d8bacefc6e/cockroach-logs4127026535
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9e2b4900e7c24e0e704b56d8bacefc6e/logTestLogic_mixed_version_udf_mutations147623490
--- FAIL: TestLogic_mixed_version_udf_mutations (120.05s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_mixed_version_udf_mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test.TestLogic_mixed_version_udf_mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12345583?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12345583?buildTab=artifacts#/) on master @ [447f59d14820752dc3703dae2371bd24f24c491a](https://github.com/cockroachdb/cockroach/commits/447f59d14820752dc3703dae2371bd24f24c491a):


```
=== RUN   TestLogic_mixed_version_udf_mutations
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/9e2b4900e7c24e0e704b56d8bacefc6e/logTestLogic_mixed_version_udf_mutations2455793794
    test_log_scope.go:81: use -show-logs to present logs inline
[12:58:49] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9739/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations: 8 statements
    logic.go:2908: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9739/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations:90: SELECT crdb_internal.is_at_least_version('23.2')
        expected:
            true
            
        but found (query options: "retry") :
            false
            
[13:00:10] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9739/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations: 14 statements
[13:00:10] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9739/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations with config cockroach-go-testserver-upgrade-to-master: 14 tests, 1 failures
    logic.go:3998: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9739/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations:95: error while processing
    logic.go:3998: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9739/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test_/cockroach-go-testserver-upgrade-to-master_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations:95: too many errors encountered, skipping the rest of the input
cockroach logs captured in: /artifacts/tmp/_tmp/9e2b4900e7c24e0e704b56d8bacefc6e/cockroach-logs3396062046
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9e2b4900e7c24e0e704b56d8bacefc6e/logTestLogic_mixed_version_udf_mutations2455793794
--- FAIL: TestLogic_mixed_version_udf_mutations (106.67s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_mixed_version_udf_mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/mixed_version_procedure
+++ b/pkg/sql/logictest/testdata/logic_test/mixed_version_procedure
@@ -33,19 +33,20 @@ CREATE PROCEDURE p() LANGUAGE SQL AS 'SELECT 1'
 
 # Upgrade all nodes.
 
-# TODO(#112621): Unskip these tests. It is currently skipped because the cluster
-# does not reliably complete the upgrade in CI.
-# upgrade 1
-#
-# upgrade 2
-#
-# # Makes sure the upgrade job has finished, and the cluster version gate is
-# # passed.
-# query B retry
-# SELECT crdb_internal.is_at_least_version('23.2')
-# ----
-# true
-#
-# # Creating a procedure should now be possible.
-# statement ok
-# CREATE PROCEDURE p() LANGUAGE SQL AS 'SELECT 1'
+upgrade 1
+
+upgrade 2
+
+statement ok
+SET CLUSTER SETTING version = crdb_internal.node_executable_version()
+
+# Makes sure the upgrade job has finished, and the cluster version gate is
+# passed.
+query B
+SELECT crdb_internal.is_at_least_version('23.2')
+----
+true
+
+# Creating a procedure should now be possible.
+statement ok
+CREATE PROCEDURE p() LANGUAGE SQL AS 'SELECT 1'

--- a/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges
+++ b/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges
@@ -71,33 +71,37 @@ SELECT crdb_internal.release_series(crdb_internal.node_executable_version())
 
 # Makes sure the upgrade job has finished, and the cluster version gate is
 # passed.
-# TODO(#112621): Unskip these tests. It is currently skipped because the cluster
-# does not reliably complete the upgrade in CI.
-# query B retry
-# SELECT crdb_internal.is_at_least_version('23.2')
-# ----
-# true
-#
-# query TTTTTTB colnames
-# SELECT * FROM [SHOW GRANTS ON FUNCTION f] ORDER BY grantee
-# ----
-# database_name  schema_name  routine_id  routine_signature  grantee  privilege_type  is_grantable
-# test           public       100106      f()                admin    ALL             true
-# test           public       100106      f()                public   EXECUTE         false
-# test           public       100106      f()                root     ALL             true
-#
-# user testuser1 nodeidx=1
-#
-# # Makes sure the user can execute f on node 1.
-# query I
-# SELECT f()
-# ----
-# 1
-#
-# user testuser1 nodeidx=2
-#
-# # Makes sure the user can execute f on node 2.
-# query I
-# SELECT f()
-# ----
-# 1
+
+user root
+
+statement ok
+SET CLUSTER SETTING version = crdb_internal.node_executable_version()
+
+query B
+SELECT crdb_internal.is_at_least_version('23.2')
+----
+true
+
+query TTTTTTB colnames
+SELECT * FROM [SHOW GRANTS ON FUNCTION f] ORDER BY grantee
+----
+database_name  schema_name  routine_id  routine_signature  grantee  privilege_type  is_grantable
+test           public       100106      f()                admin    ALL             true
+test           public       100106      f()                public   EXECUTE         false
+test           public       100106      f()                root     ALL             true
+
+user testuser1 nodeidx=1
+
+# Makes sure the user can execute f on node 1.
+query I
+SELECT f()
+----
+1
+
+user testuser1 nodeidx=2
+
+# Makes sure the user can execute f on node 2.
+query I
+SELECT f()
+----
+1

--- a/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations
+++ b/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_mutations
@@ -83,28 +83,29 @@ $$ LANGUAGE SQL;
 # Upgrade all nodes and test that creating UDFs with mutations succeeds.
 # ----------------------------------------------------------------------
 
-# TODO(#112621): Unskip these tests. It is currently skipped because the cluster
-# does not reliably complete the upgrade in CI.
-# upgrade 1
-#
-# upgrade 2
-#
-# query B retry
-# SELECT crdb_internal.is_at_least_version('23.2')
-# ----
-# true
-#
-# statement ok
-# CREATE FUNCTION f_insert() RETURNS VOID AS $$
-#   INSERT INTO t VALUES (1,1);
-# $$ LANGUAGE SQL;
-#
-# statement ok
-# CREATE FUNCTION f_delete() RETURNS VOID AS $$
-#   DELETE FROM t WHERE a = 1;
-# $$ LANGUAGE SQL;
-#
-# statement ok
-# CREATE FUNCTION f_update() RETURNS VOID AS $$
-#   UPDATE t SET b = 1 WHERE a = 1;
-# $$ LANGUAGE SQL;
+upgrade 1
+
+upgrade 2
+
+statement ok
+SET CLUSTER SETTING version = crdb_internal.node_executable_version()
+
+query B retry
+SELECT crdb_internal.is_at_least_version('23.2')
+----
+true
+
+statement ok
+CREATE FUNCTION f_insert() RETURNS VOID AS $$
+  INSERT INTO t VALUES (1,1);
+$$ LANGUAGE SQL;
+
+statement ok
+CREATE FUNCTION f_delete() RETURNS VOID AS $$
+  DELETE FROM t WHERE a = 1;
+$$ LANGUAGE SQL;
+
+statement ok
+CREATE FUNCTION f_update() RETURNS VOID AS $$
+  UPDATE t SET b = 1 WHERE a = 1;
+$$ LANGUAGE SQL;

--- a/pkg/sql/logictest/testdata/logic_test/upgrade_skip_version
+++ b/pkg/sql/logictest/testdata/logic_test/upgrade_skip_version
@@ -7,12 +7,11 @@ select crdb_internal.active_version()
 
 upgrade all
 
-# We have seen that upgrades can take a long time in CI. Give some extra time
-# for the upgrade to complete.
-retry_duration 10m
+statement ok
+SET CLUSTER SETTING version = crdb_internal.node_executable_version()
 
 # Verify that the cluster is upgrading to 24.1.
-query T retry
+query T
 SELECT crdb_internal.release_series(version) FROM [SHOW CLUSTER SETTING version]
 ----
 24.1

